storage/index.go

```diff
@@ -207,156 +205,144 @@ func newNamespaceIndexWithOptions(
-		fns.OnIndexFinalize()
-		return m3dberrors.ErrTooFuture
-	}
-
-	if !pastLimit.Before(timestamp) {
-		fns.OnIndexFinalize()
-		return m3dberrors.ErrTooPast
-	}
-
-	// Ensure metric can be converted to a valid document
-	d, err := convert.FromMetric(id, tags)
-	if err != nil {
-		fns.OnIndexFinalize()
-		return err
-	}
-
-	indexBlockStart := timestamp.Truncate(i.blockSize)
-	return i.queueWrite(indexBlockStart, d, fns)
-}
-
-func (i *nsIndex) queueWrite(
-	indexBlockStart time.Time,
-	d doc.Document,
-	fns index.OnIndexSeries,
+func (i *nsIndex) WriteBatch(
+	batch *index.WriteBatch,
 ) error {
 	i.state.RLock()
 	if !i.isOpenWithRLock() {
 		i.state.RUnlock()
 		i.metrics.InsertAfterClose.Inc(1)
-		fns.OnIndexFinalize()
-		return errDbIndexUnableToWriteClosed
+		err := errDbIndexUnableToWriteClosed
+		batch.MarkUnmarkedEntriesError(err)
+		return err
 	}
 
 	// NB(prateek): retrieving insertMode here while we have the RLock.
 	insertMode := i.state.runtimeOpts.insertMode
-	wg, err := i.state.insertQueue.Insert(indexBlockStart, d, fns)
+	wg, err := i.state.insertQueue.InsertBatch(batch)
 
 	// release the lock because we don't need it past this point.
 	i.state.RUnlock()
 
 	// if we're unable to index, we still have to finalize the reference we hold.
 	if err != nil {
-		fns.OnIndexFinalize()
+		batch.MarkUnmarkedEntriesError(err)
 		return err
 	}
 	// once the write has been queued in the indexInsertQueue, it assumes
 	// responsibility for calling the resource hooks.
 
 	// wait/terminate depending on if we are indexing synchronously or not.
 	if insertMode != index.InsertAsync {
-		// FOLLOWUP(prateek): to correctly propagate indexing error to the user in
-		// the sync case, we need a mechanism to receive notifications from
-		// the index insert queue path. Some ways to do this:
-		// (0) alloc a slice for the errors in queue, and return a wait group
-		// the slice and an index into the slice.
-		// (1) provide an error channel as input to i.insertQueue.Insert, and
-		// guarantee that receives a value on success/failure in the async
-		// insertion code path. We could eliminate the wait group if we did that.
-		// (2) we could provide an OnIndexError callback within OnIndexSeries,
-		// again ensuring it was called upon failure, and cache the error within
-		// the dbShardEntry. Then we're guaranteed that the error would be set
-		// once wg.Wait returns, and the shard insert code-paths would be able
-		// to retrieve the error from the dbShardEntry. Not pretty, but probably
-		// a lot cheaper than (1).
 		wg.Wait()
+
+		// Resort the batch by initial enqueue order
+		if numErrs := batch.NumErrs(); numErrs > 0 {
+			// Restore the sort order from whene enqueued for the caller
+			batch.SortByEnqueued()
+			return fmt.Errorf("check batch: %d insert errors", numErrs)
+		}
 	}
 
 	return nil
 }
 
-// NB: this function is called by the namespaceIndexInsertQueue.
+// WriteBatches is called by the indexInsertQueue.
```

Comment:
could you drop the FOLLOWUP note below this line, considering you just did it :D

Reply:
Sure thing.